File: 200-number-of-islands/number-of-islands.py
from collections import deque

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n, m = len(grid), len(grid[0])
        ans = 0
        vis = [[False for _ in range(m)] for _ in range (n)]
        q = deque()
        for i in range(n):
            for j in range(m):
                if vis[i][j] == False and grid[i][j] == '1':
                    q.append((i, j))
                    vis[i][j] = True
                    while q:
                        curr = q.popleft()
                        r, c = curr[0], curr[1]
                        if r > 0 and vis[r - 1][c] == False and grid[r - 1][c] == '1':
                            vis[r - 1][c] = True
                            q.append((r - 1, c))
                        if r < n - 1 and vis[r + 1][c] == False and grid[r + 1][c] == '1':
                            vis[r + 1][c] = True
                            q.append((r + 1, c))
                        if c > 0 and vis[r][c - 1] == False and grid[r][c - 1] == '1':
                            vis[r][c - 1] = True
                            q.append((r, c - 1))
                        if c < m - 1 and vis[r][c + 1] == False and grid[r][c + 1] == '1':
                            vis[r][c + 1] = True
                            q.append((r, c + 1))
                    ans += 1
        return ans

File: 200-number-of-islands/number-of-islands.java
class Solution {
    
    record State(int row, int col) {}

    public int numIslands(char[][] grid) {
        int n = grid.length, m = grid[0].length;
        int vis[][] = new int[n][m];
        int ans = 0;
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (vis[i][j] == 0 && grid[i][j] == '1') {
                    ans++;
                    Queue<State> q = new LinkedList<>();
                    q.offer(new State(i, j));
                    vis[i][j] = 1;
                    while (!q.isEmpty()) {
                        State curr = q.poll();
                        int r = curr.row(), c = curr.col();
                        int dirs[][] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
                        for (int dir[] : dirs) {
                            int new_r = r + dir[0], new_c = c + dir[1];
                            if (new_r >= 0 && new_r < n && new_c >= 0 && new_c < m && vis[new_r][new_c] == 0 && grid[new_r][new_c] == '1') {
                                q.offer(new State(new_r, new_c));
                                vis[new_r][new_c] = 1;
                            }
                        }
                    }
                }
            }
        }
        return ans;
    }
}